### PR TITLE
[WIP] Update and standardize implementation of packages, in sync with spack update

### PR DESCRIPTION
**Pull request by adrienbernede**
### Summary

This PR :

- migrates CARE and Caliper to CachedCMakePackage, reducing the gap with implementations found in llnl/radiuss-spack-configs.
- improves coherency in version constraints across RADIUSS packages.
- updates Spack.
- updates RAJA to latest develop.
- **NOTE: This PR contains changes to kernels with reductions to update to the new val-op interface which is needed to build with RAJA develop.**

👀 Deserves some consideration:
- The spec added in CI for sycl on corona had +openmp. I had to turn openmp support off due to a build error. Openmp support is turned off for the same job in RAJA CI. Then I had to turn benchmarks off. Now it's something else. Allowing the job to fail for now.

⚠️  TODO Before Merge:
- [x] Fix failing tests in intel jobs on ruby and poodle.
- [x] Wait for merge of https://github.com/spack/spack/pull/45648,
- [x] then wait for subsequent develop snapshot (Sunday).
- [x] Update Spack to new snapshot version in `.uberenv-config.json`.
- [x] Confirm CI is passing.

### CODE REVIEW

**Comment by rhornung67** [comment]
@adrienbernede I don't understand why SYCL CI is failing. I can build and run the tests manually on corona using the `scripts/lc-builds/corona_sycl.sh` script. Please note that I fixed a couple of issues in this PR: https://github.com/LLNL/RAJAPerf/pull/481.  Those changes were necessary for me to build and run successfully. 

The build script mentioned here may be helpful for you to get the CI SYCL build correct.

**Comment by rhornung67** [comment]
Also, OpenMP can be enabled for the SYCL checks.

**Comment by adrienbernede** [comment]
@rhornung67 
I had turned off OpenMP because of this:
```
CMake Error at /usr/tce/backend/installations/linux-rhel8-x86_64/gcc-10.3.1/cmake-3.23.1-mdfqd2l7c33zg7xcvqizwz25vqmp7jfw/share/cmake-3.23/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find OpenMP_C (missing: OpenMP_C_FLAGS OpenMP_C_LIB_NAMES)
Call Stack (most recent call first):
  /usr/tce/backend/installations/linux-rhel8-x86_64/gcc-10.3.1/cmake-3.23.1-mdfqd2l7c33zg7xcvqizwz25vqmp7jfw/share/cmake-3.23/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  /usr/tce/backend/installations/linux-rhel8-x86_64/gcc-10.3.1/cmake-3.23.1-mdfqd2l7c33zg7xcvqizwz25vqmp7jfw/share/cmake-3.23/Modules/FindOpenMP.cmake:545 (find_package_handle_standard_args)
  blt/cmake/thirdparty/BLTSetupOpenMP.cmake:11 (find_package)
  blt/cmake/BLTSetupTargets.cmake:86 (include)
  blt/cmake/SetupThirdParty.cmake:6 (include)
  blt/SetupBLT.cmake:129 (include)
  CMakeLists.txt:28 (include)
```
I'll try with your fix.

**Comment by rhornung67** [comment]
@adrienbernede the intel 2023 failure can be fixed by adding the flag `fp-model=precise` to the compile line (via cxxglags). Unfirtunately, that causes another check that currently passes to fail..... 😢

**Comment by adrienbernede** [comment]
@rhornung67 Is the floating point model something that can be changed for RAJA only, or should it propagate to the project using RAJA as a dependency ? In the former case, we could apply the fly via CMake or via Spack. In the latter case... well I hope it's not the latter...

**Comment by rhornung67** [comment]
Actually, all projects using the intel compiler that I know of set that flag. However, we should apply it via CMake and not Spack.

**Review by rhornung67 — APPROVED** [review]

**Comment by rhornung67** [comment]
@adrienbernede it looks like the SYCL job on corona is not running on a node with GPUs. Will you please take a look? Thank you.

**Comment by rhornung67** [comment]
**Don't merge this until SYCL check on corona passes.**

**Review by rhornung67 — COMMENTED** [review]
Let's not merge this until we fix the corona SYCL CI job. I haven't been able to reproduce the failure.

**Comment by rhornung67** [comment]
@adrienbernede I pushed a change to the raja-perf spack package that may fix the SYCL CI issue. If this works, we will need to merge a sequence of PRs, in order: first pull radiuss-spack-configs branch with Spack package change, second pull updated radiuss-spack-configs main into RAJA develop, third pull new RAJA develop into this PR branch. Then, we should be able to merge this (fingers crossed).

**Comment by rhornung67** [comment]
@adrienbernede I found and fixed the issue. See this commit: [01bf5b2](https://github.com/LLNL/RAJAPerf/pull/480/commits/01bf5b28fa0f2d916a72f4f4ae16b3af46df1a7f)

**Comment by rhornung67** [comment]
@rchen20 @MrBurmark This will be merged soon after a couple of other PRs are merged on other projects so that this can point at develop/main branches of RAJA and radiuss-spack-configs. To do these updates, we pulled the reduction kernels changes to the new val-op interface into this PR. Please review those changes in this PR before this is merged. Thank you.

File: src/lcals/FIRST_MIN-Cuda.cpp
Comment (rchen20) [inline src/lcals/FIRST_MIN-Cuda.cpp:181]: The `ValLocOp` `Index_type` should match the above `ValLoc`'s `RAJA::Index_type`. I can't quite tell if they match, but we should check. It's probably working due to type erasure in the internals.

File: src/lcals/FIRST_MIN-Sycl.cpp
Comment (rchen20) [inline src/lcals/FIRST_MIN-Sycl.cpp:98]: Should check matching `Index_type` here too.

File: src/lcals/FIRST_MIN-Seq.cpp
Comment (rchen20) [inline src/lcals/FIRST_MIN-Seq.cpp:114]: Should check matching `Index_type` here too.

File: src/lcals/FIRST_MIN-Cuda.cpp
Comment (rhornung67) [reply src/lcals/FIRST_MIN-Cuda.cpp]: Good catch -- thanks!. That's a cut-and-paste error.  The types are the same and the types in RAJAPerf are aliases to the RAJA types. I pushed a change plus some others for code consistency.

File: src/lcals/FIRST_MIN-Sycl.cpp
Comment (rhornung67) [reply src/lcals/FIRST_MIN-Sycl.cpp:98]: Fixed.

File: src/lcals/FIRST_MIN-Seq.cpp
Comment (rhornung67) [reply src/lcals/FIRST_MIN-Seq.cpp:115]: Fixed.

**Review by rchen20 — APPROVED** [review]
LGTM

**Comment by rhornung67** [comment]
@adrienbernede This PR is synced up with latest radiuss-spack-configs main and RAJA develop. When it passes checks, it is good to merge.

**Review by rhornung67 — APPROVED** [review]

**Pull request merged**
